pkg/controllers/rollout/controller.go

```diff
@@ -0,0 +1,446 @@
+	klog.V(2).InfoS("Calculated the max number of bindings to remove", "clusterResourcePlacement", klog.KObj(crp),
+		"maxUnavailableNumber", maxUnavailableNumber, "minAvailableNumber", minAvailableNumber,
+		"lowerBoundAvailableBindings", lowerBoundAvailableNumber, "maxNumberOfBindingsToRemove", maxNumberToRemove)
+	if maxNumberToRemove > 0 {
+		i := 0
+		// we first remove the bindings that are not selected by the scheduler anymore
+		for ; i < maxNumberToRemove && i < len(removeCandidates); i++ {
+			tobeUpdatedBinding = append(tobeUpdatedBinding, removeCandidates[i])
+		}
+		// we then update the bound bindings to the latest resource resourceBinding which will lead them to be unavailable for a short period of time
+		j := 0
+		for ; i < maxNumberToRemove && j < len(updateCandidates); i++ {
+			tobeUpdatedBinding = append(tobeUpdatedBinding, updateCandidates[j])
+			j++
+		}
+	}
+
+	// calculate the max number of bindings that can be added according to user specified MaxSurge
+	maxSurgeNumber, _ := intstr.GetScaledValueFromIntOrPercent(crp.Spec.Strategy.RollingUpdate.MaxSurge, targetNumber, true)
+	maxReadyNumber := targetNumber + maxSurgeNumber
+	// This is the upper bound of the number of bindings that can be ready during the rolling update
+	// We count anything that still has work object on the hub cluster as can be ready since the member agent may have connection issue with the hub cluster
+	upperBoundReadyNumber := len(canBeReadyBindings)
+	maxNumberToAdd := maxReadyNumber - upperBoundReadyNumber
+
+	klog.V(2).InfoS("Calculated the max number of bindings to add", "clusterResourcePlacement", klog.KObj(crp),
+		"maxSurgeNumber", maxSurgeNumber, "maxReadyNumber", maxReadyNumber, "upperBoundReadyBindings",
+		upperBoundReadyNumber, "maxNumberOfBindingsToAdd", maxNumberToAdd)
+	for i := 0; i < maxNumberToAdd && i < len(boundingCandidates); i++ {
+		tobeUpdatedBinding = append(tobeUpdatedBinding, boundingCandidates[i])
+	}
+
+	return tobeUpdatedBinding, true
+}
+
+// isBindingReady checks if a binding is considered ready.
+// A binding is considered ready if the binding's current spec has been applied before the ready cutoff time.
+func isBindingReady(binding *fleetv1beta1.ClusterResourceBinding, readyTimeCutOff time.Time) bool {
+	// find the latest applied condition that has the same generation as the binding
+	appliedCondition := binding.GetCondition(string(fleetv1beta1.ResourceBindingApplied))
+	if appliedCondition != nil && appliedCondition.Status == metav1.ConditionTrue &&
+		appliedCondition.ObservedGeneration == binding.GetGeneration() {
+		return appliedCondition.LastTransitionTime.Time.Before(readyTimeCutOff)
+	}
+	return false
+}
+
+// updateBindings updates the bindings according to its state
+func (r *Reconciler) updateBindings(ctx context.Context, latestResourceSnapshotName string, tobeUpgradedBinding []*fleetv1beta1.ClusterResourceBinding) error {
+	// issue all the update requests in parallel
```

Comment:
doesn't the for loop kinda check that already?

Reply:
The functionality is the same.
adding a shortcut could skip creating the errs. 

I don't have strong opinion on this and it may be a "niiiiit" comment. :)